skt/publisher.py

```diff
@@ -17,134 +17,101 @@
-    TYPE = 'sftp'
-
-    def publish(self, source):
+        if self.pub_type == 'cp':
+            self.cp()
+        elif self.pub_type == 'scp':
+            self.scp()
+        else:
+            raise(
+                KeyError,
+                "Publisher type not supported: {}".format(self.pub_type)
+            )
+
+        return self.geturl()
+
+    def geturl(self):
         """
-        Copy the source file to public destination.
+        Get published URL for a source file path.
 
         Args:
             source: Source file path.
 
         Returns:
             Published URL corresponding to the specified source.
         """
-        proc = subprocess.Popen(['sftp', self.destination],
-                                stdin=subprocess.PIPE)
-        proc.stdin.write("put -r %s\n" % source)
-        proc.stdin.close()
-        proc.wait()
-        return self.geturl(source)
-
-
-def getpublisher(ptype, parg, pburl):
-    """
-    Create an instance of a "publisher" subclass with specified arguments.
-
-    Args:
-        rtype:  The value of the class "TYPE" member to match.
-        rarg:   A dictionary with the instance creation arguments.
-
-    Returns:
-        The created class instance.
-
-    Raises:
-        ValueError if the rtype match wasn't found.
-    """
-    for cls in Publisher.__subclasses__():
-        if cls.TYPE == ptype:
-            return cls(parg, pburl)
-    raise ValueError("Unknown publisher type: %s" % ptype)
+        return join_with_slash(self.baseurl, os.path.basename(self.source))
+
+    def cp(self):
+        """Copy the source file to public destination."""
+        # If the destination directory doesn't exist already, create it.
+        if not os.path.isdir(self.destination):
+            logging.debug("Creating destination: %s", self.destination)
+            os.makedirs(self.destination)
+
+        # Files can be copied directly from the source to destination.
+        if os.path.isfile(self.source):
+            logging.debug(
+                "copying file: %s > %s",
+                self.source,
+                self.destination
+            )
+            shutil.copy(self.source, self.destination)
+
+        # Directories (RPM repositories) require a tree to be copied.
+        if os.path.isdir(self.source):
+            logging.debug(
+                "copying tree: %s > %s",
+                self.source,
+                self.destination
+            )
+            # Add the directory name to the destination so that copytree will
+            # copy files.
+            destination = join_with_slash(
+                self.destination,
+                os.path.basename(self.source)
+            )
+
+            # copytree will not copy to a directory that exists.
+            if os.path.isdir(destination):
+                shutil.rmtree(destination)
+
+            shutil.copytree(self.source, destination)
+
+    def scp(self):
```

Comment:
You need to pass `-r` here for handling folders. Actually, if you add `-r` scp can handle local copies as well which means you can remove `def cp` and just use `scp` for local and remote copies :man_shrugging:

Reply:
Nice find. I've updated the patch.